src/migrateWidgetsWithinDashboard.ts

```diff
@@ -0,0 +1,67 @@
+import { DataModel, Layout, getLayoutPath } from "@activeviam/activeui-sdk-5.1";
+import { _removeWidgetFromPage } from "@activeviam/dashboard-base-5.1";
+import _forEach from "lodash/forEach";
+import { MigrateWidgetCallback } from "./migration.types";
+import { WidgetFlaggedForRemovalError } from "./WidgetFlaggedForRemovalError";
+
+/**
+ * Mutates `dashboardState` by calling `callback` on each underlying widget state.
+ */
+export function migrateWidgetsWithinDashboard<
+  FromWidgetState extends { widgetKey: string; name: string },
+  FromDashboardState extends {
+    pages: {
+      [pageKey: string]: {
+        name: string;
```

Comment:
Could we use `DashboardPageState` here?

Reply:
We could, and same for `DashboardState` and `WidgetState`, but it would tie it down to a specific version of the type (e.g. the 5.0 one).
Here I wanted to avoid that since `migrateWidgetsWithinDashboard` will be called by all `migrateXToY` functions.
And the content type is not needed to implement the iteration logic: just this shell is enough.

It's not a very strong preference though, so if you prefer to import from `5.0` and try to type more precisely I don't mind giving it a go.